File: apps-microservices/graph-rag-api-recherche-service/app/services/product_service.py
import logging
from typing import List, Optional, Any
from app.domain.models import CaracteristiqueResponse
from app.infrastructure.clients import clients


class ProductService:
    async def get_characteristics(
        self, product_id: str
    ) -> Optional[List[CaracteristiqueResponse]]:
        """
        Retrieves the technical characteristics for a specific product by its Business ID (id_produit).
        Only returns characteristics that have a non-null/non-empty value (scalar or range).
        """
        cypher_query = """
        MATCH (p:Produit {id_produit: $pid})
        OPTIONAL MATCH (p)-[:A_POUR_CARACTERISTIQUE]->(c:CaracteristiqueTechnique)
        WHERE (c.valeur IS NOT NULL AND toString(c.valeur) <> "") 
           OR c.valeur_min IS NOT NULL 
           OR c.valeur_max IS NOT NULL
        RETURN p, collect(c {.*}) as characteristics
        """

        try:
            results = await clients.execute_cypher(cypher_query, {"pid": product_id})

            if not results:
                logging.info(f"Product with ID '{product_id}' not found.")
                return None

            record = results[0]
            raw_chars = record.get("characteristics", [])

            characteristics = []
            for raw in raw_chars:
                try:
                    if not raw.get("nom") or not raw.get("label"):
                        continue

                    char = CaracteristiqueResponse(
                        nom=raw.get("nom"),
                        label=raw.get("label"),
                        valeur=raw.get("valeur"),
                        unite=raw.get("unite"),
                        type_donnee=raw.get("type_donnee"),
                        valeur_min=raw.get("valeur_min"),
                        valeur_max=raw.get("valeur_max"),
                        id_caracteristique=str(raw.get("id_source_caracteristique")),
                        id_valeur=str(raw.get("id_source_valeur", "N/A")),
                    )
                    characteristics.append(char)
                except Exception as e:
                    logging.warning(
                        f"Failed to parse characteristic for product {product_id}: {e}"
                    )
                    continue

            # Sort by ID if possible, else keep order
            characteristics.sort(
                key=lambda x: (
                    x.id_caracteristique
                    if x.id_caracteristique.isdigit()
                    else x.id_caracteristique
                )
            )
            return characteristics

        except Exception as e:
            logging.error(
                f"Error retrieving characteristics for product {product_id}: {e}",
                exc_info=True,
            )
            raise e

    async def delete_produit(self, product_id: str) -> Optional[dict]:
        """
        Deletes a product node by its ID and returns the deleted node's properties.
        """
        cypher_query = """
        MATCH (p:Produit {id_produit: $pid})
        WITH p, properties(p) as props
        DETACH DELETE p
        RETURN props
        """

        try:
            results = await clients.execute_cypher(cypher_query, {"pid": product_id})

            if not results:
                logging.info(f"Product with ID '{product_id}' not found for deletion.")
                return None

            return results[0].get("props")

        except Exception as e:
            logging.error(
                f"Error deleting product {product_id}: {e}",
                exc_info=True,
            )
            raise e


product_service = ProductService()

File: apps-microservices/graph-rag-api-recherche-rust-service/src/services/product_service.rs
use serde_json::{json, Value};
use tracing::{error, info};

use crate::infrastructure::clients::CLIENTS;

/// Product service for characteristics and deletion, mirroring Python's product_service.py.
pub struct ProductService;

impl ProductService {
    /// Get characteristics for a product by ID.
    pub async fn get_caracteristiques(id_produit: &str) -> Vec<Value> {
        let query = r#"
            MATCH (p:Produit {id_produit: $id_produit})-[:A_POUR_CARACTERISTIQUE]->(c:CaracteristiqueTechnique)
            RETURN c {
                .id_source_caracteristique,
                .label,
                .valeur,
                .valeur_min,
                .valeur_max,
                .unite,
                .unite_canonique,
                .valeur_canonique,
                .valeur_min_canonique,
                .valeur_max_canonique,
                .type_donnee,
                .id_source_valeur
            } AS caracteristique
            ORDER BY c.label
        "#;

        let formatted_id = if id_produit.starts_with("id_produit_") {
            id_produit.to_string()
        } else {
            format!("id_produit_{}", id_produit)
        };

        let params = json!({ "id_produit": formatted_id });
        let results = CLIENTS.execute_cypher(query, &params).await;

        results
            .into_iter()
            .filter_map(|r| r.get("caracteristique").cloned())
            .collect()
    }

    /// Delete a product and its relationships.
    pub async fn delete_product(id_produit: &str) -> (bool, String) {
        let formatted_id = if id_produit.starts_with("id_produit_") {
            id_produit.to_string()
        } else {
            format!("id_produit_{}", id_produit)
        };

        let query = r#"
            MATCH (p:Produit {id_produit: $id_produit})
            OPTIONAL MATCH (p)-[r]-()
            DELETE r, p
            RETURN count(p) AS deleted_count
        "#;

        let params = json!({ "id_produit": formatted_id });

        match CLIENTS.execute_cypher(query, &params).await.first() {
            Some(result) => {
                let count = result
                    .get("deleted_count")
                    .and_then(|v| v.as_i64())
                    .unwrap_or(0);
                if count > 0 {
                    (true, format!("Product {} deleted successfully", id_produit))
                } else {
                    (false, format!("Product {} not found", id_produit))
                }
            }
            None => (false, "Failed to execute delete query".to_string()),
        }
    }
}
